Namespace scoped resource clients and watch improvements

marcogschmidt opened the pull request:
This PR includes two main (non-breaking) changes to the Kubernetes resource clients:

**1 - Resource clients can now be namespace-scoped**
The `KubeResourceClientFactory` can now be configured with a namespace whitelist. When the client is `Register`ed, it will create a `SharedIndexedInformer` for each of the namespaces in this list (more specifically, each of the informer `ListWatch`es will target only one namespace). Previously, only one informer was created and it would list/watch all namespaces. The problem with this was that it required the client to have cluster-wide access to resources. Now we can build resource clients for users that have access to only a limited number of namespaces. E.g. if we have a user that can only access `ChangeSets` in the `gloo-system` namespace we can do:
```go
csClient, err := v1.NewChangeSetClientWithToken(&factory.KubeResourceClientFactory{
		...
		NamespaceWhitelist:      []string{"gloo-system"},
	}, <token_that_only_has_access_to_changesets_in_gloo-system>)
```

**2 - Resource client watches will only receive events for the relevant resource/namespace**
A new `ResourceEventHandler` now provides more context as to which resource was actually updates. The `ResourceClient` uses the additional information to filter out events that are not relevant to the specific watch. Earlier any event on the shared cache would notify all the watches created by clients that shared that cache. Now e.g. a watch created for the `csClient` in the example above would be notified only when an event occurred on some `ChangeSet` in the `gloo-system` namespace.

**Other minor changes**
- a `KubeResourceClientFactory` can now specify its own `resyncPeriod` for informer cache refreshes
- added a 10s timeout when calling the `KubeController.Run(...)` method in order to avoid blocking the whole process if the initial informer cache synchronization cannot be performed (e.g. if the client does not have access to a resource)

**NOTE:** All changes are backwards compatible:
- if no `namespaces` list is specified on the `KubeResourceClientFactory`, we default to all namespaces
- if no `resyncPeriod` is specified on the `KubeResourceClientFactory`, we default to the `SharedInformerFactory` resync period (12h, as before)

marcogschmidt commented on pkg/api/v1/clients/kube/resource_client.go:
```diff
@@ -378,29 +411,42 @@ func (rc *ResourceClient) Watch(namespace string, opts clients.WatchOpts) (<-cha
 
 		for {
 			select {
 			case <-time.After(opts.RefreshRate): // TODO(yuval-k): can we remove this? is the factory takes care of that...
```
@yuval-k I would remove it. This does not really refresh the cache, but merely re-lists its content. The actual resync period (that forces the reflector to check whether the caches are out of sync) can now be specified on the `KubeResourceClientFactory`. Not sure if this is needed somewhere though. In case, the same behavior can be achieved by periodically calling the `List` method on the client itself. What do you think?

ilackarms replied:
the idea is that this helps "kick the tires" to make sure event loops get started immediately upon calling a watch. if you noticed this function isn't necessary, maybe because a watch event happens immediately anyway, we could remove it. but i think removing it might disturb some of our event loops (in particular i believe function discovery benefitted from this)

marcogschmidt replied:
I think you might be referring to line 420? I was talking about the `TODO` @yuval-k left at line 424 (the timeout clause in the `select` statement).

marcogschmidt commented on pkg/api/v1/clients/kube/controller/controller.go:
```diff
@@ -17,132 +17,112 @@ import (
 	informers ...cache.SharedInformer) *Controller {
 
+	// TODO: how will this work with unauthorized users?
 	eventBroadcaster := record.NewBroadcaster()
```
@ilackarms @yuval-k It looks like this `recorder` is currently not used. Right now the controller works and I did not investigate further. I can see how this would fail in case `kubeclient` does not have permissions on `Event`s. What do you think?

marcogschmidt replied:
Ok, will do

marcogschmidt commented on pkg/api/v1/clients/kube/resource_client_factory.go:
```diff
@@ -64,137 +67,174 @@ type ResourceClientSharedInformerFactory struct {
-		f.defaultResync, // TODO(yuval-k): make this configurable!
-		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
-	)
+			&v1.Resource{}, // TODO(yuval-k): can we make this rc.crd.Type ?
```
@yuval-k Not without other changes. Tried just to see if it would work and got this error:
```
ERROR: logging before flag.Parse: E1231 15:42:13.213211    6035 reflector.go:315] github.com/solo-io/solo-kit/pkg/api/v1/clients/kube/controller/controller.go:88: expected type *v1.ChangeSet, but watch event object had type *v1.Resource
```
Thrown from [here](https://github.com/kubernetes/client-go/blob/7856fdbcc3be4d133402982f0a5069d251eb8459/tools/cache/reflector.go#L315).

ilackarms replied:
can you explain a bit more? is this still an issue?

marcogschmidt replied:
I was replying to the second `TODO` comment. The parameter is used by the reflector to check the type of the event object it receives from the `ListWatch`. By passing `&v1.Resource{}` it works, but if you pass `rc.crd.Type` you get the error I mentioned. 

I believe the parameter must match the resource type that the `WatchFunc` in line 121
```go
watch := rc.kube.ResourcesV1().Resources(ns).Watch
```
is defined for, if that makes sense. Yuval was probably wondering if by passing `rc.crd.Type` you could get type information with the event for free, but from what I was able to tell there is no way of getting anything more than our `*v1.Resource`.

ilackarms commented on pkg/api/v1/clients/factory/resource_client_factory.go:
```diff
@@ -50,10 +59,19 @@ func newResourceClient(factory ResourceClientFactory, params NewResourceClientPa
+		// 1. If no namespace list was provided, default to all namespaces
+		// 2. Error if namespace list contains the empty string plus other values
+		namespaces := opts.Namespaces
+		if namespaces == nil || len(namespaces) == 0 {
```
nit: `len(namespaces) == 0` covers the `namespaces == nil`

https://golang.org/pkg/builtin/#len

ilackarms commented on pkg/api/v1/clients/factory/resource_client_factory.go:
```diff
@@ -50,10 +59,19 @@ func newResourceClient(factory ResourceClientFactory, params NewResourceClientPa
+		namespaces := opts.Namespaces
+		if namespaces == nil || len(namespaces) == 0 {
+			namespaces = []string{metaV1.NamespaceAll}
+		} else if stringutils.ContainsString(metaV1.NamespaceAll, namespaces) && len(namespaces) > 1 {
```
nit: remove `else`, change to 
`if stringutils.ContainsString(metaV1.NamespaceAll, namespaces) {` 
(if `len(namespaces) == 0` then `stringutils.ContainsString(metaV1.NamespaceAll, namespaces)` will always be false)

marcogschmidt replied:
Will change it

ilackarms commented on pkg/api/v1/clients/factory/resource_client_factory.go:
```diff
@@ -81,11 +99,15 @@ type ResourceClientFactory interface {
 	Cfg             *rest.Config
 	SharedCache     *kube.KubeCache
 	SkipCrdCreation bool
+	Namespaces      []string
```
suggestion: replace `Namespaces` with something like `NamespaceWhitelist` so it's more clear what function the list serves

marcogschmidt replied:
Good point, will do

ilackarms commented on pkg/api/v1/clients/kube/resource_client_factory.go:
```diff
@@ -64,137 +67,174 @@ type ResourceClientSharedInformerFactory struct {
 
-	informers map[reflect.Type]cache.SharedIndexInformer
-	// startedInformers is used for tracking which informers have been started.
+	informers map[reflect.Type]map[string]cache.SharedIndexInformer
```
maybe consider converting this `map[reflect.Type]map[string]cache.SharedIndexInformer` to some kind of struct for easier management? it''s not terrible, but having a struct enclosing these maps with some `Add` `Remove` functions can simplify this a bit

marcogschmidt replied:
Makes a lot of sense, I'll do it

ilackarms commented:
overall looks good, i gave some style suggestions 

my only concern is that this is a change to a fairly complicated (and not well-tested) piece of the code. how hard do you think it would be to add some test coverage around this code?

marcogschmidt commented on pkg/api/v1/clients/kube/controller/controller.go:
```diff
@@ -17,132 +17,112 @@ import (
+//
+// When a message is received on the given stopCh, the controller will stop the informers, shutdown the work queue and
+// wait for workers to finish processing their current work items.
+func (c *Controller) Run(parallelism int, stopCh <-chan struct{}) error {
```
😊

marcogschmidt commented on pkg/api/v1/clients/kube/controller/controller.go:
```diff
@@ -17,132 +17,112 @@ import (
 
-// NewController returns a new controller
+// Returns a new kubernetes controller without starting it. The given event handler is registered with each of the
+// given informers. The controller also collects the HasSynced functions of each of the informers.
```
Good catch, will update the comment

marcogschmidt commented on pkg/api/v1/clients/kube/controller/controller.go:
```diff
@@ -17,132 +17,112 @@ import (
+		defer c.workQueue.ShutDown()
+		log.Debugf("Starting workers")
+
+		// Launch two workers to process resources
```
Fixed

marcogschmidt commented on pkg/api/v1/clients/kube/resource_client.go:
```diff
@@ -162,13 +159,18 @@ type ResourceClient struct {
 	if err != nil {
 		return nil, errors.Wrapf(err, "creating api extensions client")
 	}
+	//
```
Done

marcogschmidt commented:
Summary of the commits since the last review: 
- added tests fo
  - k8s controller
  - shared cache
  - kube resource client
- simplified shared cache design and resolved bugs that were surfaced by the tests
- configured CI to run all test suites
  - NOTE: I had to comment out a lot of tests since they were failing and fixing them is beyond the scope of this PR

ilackarms commented on pkg/api/v1/clients/kube/resource_client.go:
```diff
@@ -311,16 +248,23 @@ func (rc *ResourceClient) Delete(namespace, name string, opts clients.DeleteOpts
+	}
+
+	// Will have no effect if the factory is already running
+	rc.sharedCache.Start(context.TODO())
```
do we want to use `opts.WatchOpts.Ctx` here? @yuval-k

yuval-k replied:
In theory, as long as the credentials do not change the cache neednt expire.
Assuming opts.WatchOpts.Ctx expires on settings change which should be rare, i don't see a big different either way for performance. the question is of correctness:

is Start be called again when that context is canceled? what is the lifetime of sharedCache?

marcogschmidt replied:
That's a good point. Right now the controller (informers + workers) is stopped when the context that you pass to `Start(ctx)` is cancelled., which is never, since we are passing a todo context.

The most correct behavior would probably be to get the context in the resourceClient constructor and to propagate it to the controller. That way client code can determine their own lifecycle for the resourceClient.

yuval-k replied:
it's fine that they have never stopped, as long as it is consistent - 
i.e. when the next start is called it will re-use the existing object?

marcogschmidt replied:
After you call `start` the first time, subsequent invocations do not have any effect, and you can't register clients after the cache has been started. So there is no way to reuse it. You have to register all the clients that want to use the cache beforehand and then start it. After that it just lives forever. 

You're right that it's probably not a problem, given that AFAIK in our applications the client lifetime is the same as the one of the application, so there's no risk of leaking resources. Less than ideal, but good for now I guess.

marcogschmidt replied:
Maybe we could add a second constructor (not to break backwards-compatibility) in the form:
```go
NewKubeCache(done <-chan struct{})
```
That would be pretty quick.

yuval-k replied:
i would worry about backward-compatibility as there is no stable release.
lets talk about tomorrow morning - as I'm trying to understand why it doesn't work now vs before the change.

marcogschmidt replied:
Nothing really changed since the prior version, we already were passing a `ctx.TODO()` to the factory and never shutting it down (see [current `master`](https://github.com/solo-io/solo-kit/blob/bd125b289e404433f59b32148cb6eb54c488c5ec/pkg/api/v1/clients/kube/resource_client.go#L321)).

ilackarms commented on pkg/api/v1/clients/kube/resource_client.go:
```diff
@@ -356,10 +300,15 @@ func (rc *ResourceClient) List(namespace string, opts clients.ListOpts) (resourc
+		return nil, nil, err
+	}
+
+	rc.sharedCache.Start(context.TODO())
```
do we want to use opts.WatchOpts.Ctx here? @yuval-k

marcogschmidt commented on pkg/api/v1/clients/kube/resource_client_factory.go:
```diff
@@ -54,145 +56,248 @@ func init() {
+			case <-time.After(f.watchTimeout):
+				contextutils.LoggerFrom(context.TODO()).Errorf("timed out after waiting for %v"+
+					"for watch to receive event on resource %v", f.watchTimeout, resource)
+			}
```
Any comment about this? @ilackarms @yuval-k
